crates/astria-sequencer/src/app/vote_extension.rs

```diff
@@ -0,0 +1,571 @@
+        extended_commit_info: &ExtendedCommitInfo,
+    ) -> anyhow::Result<()> {
+        if height == 1 {
+            // we're processing block 1, so nothing to validate (no last commit yet)
+            return Ok(());
+        }
+
+        // inside process_proposal, we must validate the vote extensions proposed against the last
+        // commit proposed
+        validate_extended_commit_against_last_commit(last_commit, extended_commit_info)?;
+
+        validate_vote_extensions(state, height, extended_commit_info)
+            .await
+            .context("failed to validate vote extensions in validate_extended_commit_info")?;
+        Ok(())
+    }
+}
+
+// see https://github.com/skip-mev/slinky/blob/5b07f91d6c0110e617efda3f298f147a31da0f25/abci/ve/utils.go#L111
+async fn validate_vote_extensions<S: StateReadExt>(
+    state: &S,
+    height: u64,
+    extended_commit_info: &ExtendedCommitInfo,
+) -> anyhow::Result<()> {
+    use tendermint_proto::v0_38::types::CanonicalVoteExtension;
+
+    let chain_id = state
+        .get_chain_id()
+        .await
+        .context("failed to get chain id")?;
+
+    // total validator voting power
+    let mut total_voting_power: u64 = 0;
+    // the total voting power of all validators which submitted vote extensions
+    let mut submitted_voting_power: u64 = 0;
+
+    let validator_set = state
+        .get_validator_set()
+        .await
+        .context("failed to get validator set")?;
+
+    for vote in &extended_commit_info.votes {
+        total_voting_power = total_voting_power.saturating_add(vote.validator.power.value());
+
+        if vote.sig_info == Flag(tendermint::block::BlockIdFlag::Commit)
+            && vote.extension_signature.is_none()
+        {
+            anyhow::bail!(
+                "vote extension signature is missing for validator {}",
+                crate::address::base_prefixed(vote.validator.address)
+            );
+        }
+
+        if vote.sig_info != Flag(tendermint::block::BlockIdFlag::Commit)
+            && !vote.vote_extension.is_empty()
+        {
+            anyhow::bail!(
+                "non-commit vote extension present for validator {}",
+                crate::address::base_prefixed(vote.validator.address)
+            );
+        }
+
+        if vote.sig_info != Flag(tendermint::block::BlockIdFlag::Commit)
+            && vote.extension_signature.is_some()
+        {
+            anyhow::bail!(
+                "non-commit extension signature present for validator {}",
+                crate::address::base_prefixed(vote.validator.address)
+            );
+        }
+
+        if vote.sig_info != Flag(tendermint::block::BlockIdFlag::Commit) {
+            continue;
+        }
+
+        submitted_voting_power =
+            submitted_voting_power.saturating_add(vote.validator.power.value());
+
+        let verification_key = validator_set
+            .get(vote.validator.address)
+            .context("validator not found")?
```

Comment:
this should error imo, as in the process_proposal phase it means the proposer proposed an invalid commit. in the prepare_proposal phase, this shouldn't error as we got the extended commit from our cometbft node, so that would mean there's a cometbft bug